### ISSUE TITLE
Error when trying to run .pb.rb compiled by ruby-protobuf

### ISSUE THREAD

**Issue by henriklb** [logstash-plugins/logstash-codec-protobuf#28]
I compiled my .proto files with https://github.com/ruby-protobuf/protobuf and NOT the suggested complier because I need the extensions feature. When using the plugin I get the error below. 

I am new to this, I would appreciate your help. 

[optics.pb.rb.txt](https://github.com/logstash-plugins/logstash-codec-protobuf/files/2163698/optics.pb.rb.txt)
[optics.proto.txt](https://github.com/logstash-plugins/logstash-codec-protobuf/files/2163699/optics.proto.txt)
[telemetry_top.pb.rb.txt](https://github.com/logstash-plugins/logstash-codec-protobuf/files/2163700/telemetry_top.pb.rb.txt)
[telemetry_top.proto.txt](https://github.com/logstash-plugins/logstash-codec-protobuf/files/2163701/telemetry_top.proto.txt)

When I try to run my complied files I get this error message :

> [ERROR] 2018-07-04 16:57:39.517 [Converge PipelineAction::Create<main>] agent - Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"LoadError", :message=>"no such file to load -- protobuf", :backtrace=>["org/jruby/RubyKernel.java:955:in `require'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'", "/root/protoFiles/telemetry_top.pb.rb:6:in `<main>'", "org/jruby/RubyKernel.java:955:in `require'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-codec-protobuf-1.1.0/lib/logstash/codecs/protobuf.rb:393:in `load_protobuf_definition'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-codec-protobuf-1.1.0/lib/logstash/codecs/protobuf.rb:99:in `block in register'", "org/jruby/RubyArray.java:1734:in `each'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-codec-protobuf-1.1.0/lib/logstash/codecs/protobuf.rb:99:in `register'", "/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:20:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/plugins/plugin_factory.rb:97:in `plugin'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:110:in `plugin'", "(eval):1:in `<eval>'", "org/jruby/RubyKernel.java:994:in `eval'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:82:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:167:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline_action/create.rb:40:in `execute'", "/usr/share/logstash/logstash-core/lib/logstash/agent.rb:305:in `block in converge_state'"]}

**Comment by IngaFeick** [comment]
Hi, please provide your logstash config, especially the input section. Thank you.

**Comment by IngaFeick** [comment]
Also: which logstash version are you using and which version of the plugin did you install? Thanks.

**Comment by henriklb** [comment]
[logstash_conf.conf.txt](https://github.com/logstash-plugins/logstash-codec-protobuf/files/2163710/logstash_conf.conf.txt)
Logstash 6.3.0
logstash-codec-protobuf 1.1.0

**Comment by henriklb** [comment]
If I only compile the _single_ file telemetry_top.proto **with** the [recommenced compiler](https://github.com/codekitchen/ruby-protocol-buffers)  that works just fine. However, then I'm not able to decode the entire message (missing optics.proto)

As already mentioned in my first post, when I try with an alternative compiler then it will not run at all.

**Comment by IngaFeick** [comment]
These things jump at me: 

- the telemetry_top.pb.rb is included in your logstash config but it requires another pb file called 'google/protobuf/descriptor.pb'. What's the deal with that one? Does it exist? And why is its file extension not .rb ?
- you only provided one file in the logstash config but it should be two, because the telemetry_top one is being required by the optics file. All required ruby protobuf files need to be listed in the include path, in a top to bottom order (if you think about the document hierarchy). So you'd have to change that to at least this: 
`include_path => ['/root/protoFiles/telemetry_top.pb.rb', '/root/protoFiles/optics.pb.rb']` 
- your class name might be wrong. Since the optics file includes the telemetry file, the natural assumption would be that this class is the upper-/outermost protobuf definition and should be in the class name config like so:
`class_name=> "Optics"`
but I don't see how the Telemetry class is actually being used inside the Optics class, aside from the telemetry_options stuff, so I am not sure which advice to give you here. Can you check which one of those two actually goes out onto the wire and if you really need both of them?
- what kind of extension feature is it that you are using? Can you direct me to some documentation for the `:".telemetry_options" => { :is_key => true }` stuff? Thanks.

**Comment by henriklb** [comment]
1. In my current environment it does not exist. It has been used [here](https://github.com/JNPRAutomate/fluent-plugin-juniper-telemetry/blob/master/junos-telemetry/google/protobuf/descriptor.proto) in a fluentd plugin . I do not think it is needed the import of 'google/protobuf/descriptor.pb' is needed in any way.

2. I tried both with a one file and both files in the` include_path` , still very similar error message.

> [ERROR] 2018-07-04 18:07:39.173 [Converge PipelineAction::Create<main>] agent - Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"LoadError", :message=>"no such file to load -- protobuf", :backtrace=>["org/jruby/RubyKernel.java:955:in `require'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'", "/root/protoFiles/telemetry_top.pb.rb:6:in `<main>'", "org/jruby/RubyKernel.java:955:in `require'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-codec-protobuf-1.1.0/lib/logstash/codecs/protobuf.rb:393:in `load_protobuf_definition'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-codec-protobuf-1.1.0/lib/logstash/codecs/protobuf.rb:99:in `block in register'", "org/jruby/RubyArray.java:1734:in `each'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-codec-protobuf-1.1.0/lib/logstash/codecs/protobuf.rb:99:in `register'", "/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:20:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/plugins/plugin_factory.rb:97:in `plugin'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:110:in `plugin'", "(eval):1:in `<eval>'", "org/jruby/RubyKernel.java:994:in `eval'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:82:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:167:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline_action/create.rb:40:in `execute'", "/usr/share/logstash/logstash-core/lib/logstash/agent.rb:305:in `block in converge_state'"]} 

3. These files decoded worked fine with only these two file when I captured traffic with netcat and issued the following command on a UNIX machine: `protoc --decode TelemetryStream optics.proto  < data.gpb`
The only to files in needed and in the same directory were 'optics.proto' and  'telemetry_top.proto'
[step by step how I managed to decode the gpb data](https://www.juniper.net/documentation/en_US/junos/topics/reference/general/junos-telemetry-interface-decoding-data.html)

4. Full disclosure, I didn't create these .proto files myself, so I have no idea.  They were provided by Juniper Networks to use with their telemetry data streaming API.

**Comment by henriklb** [comment]
I noticed that because I use the https://github.com/ruby-protobuf/protobuf compiler the complied '.pb.rb' files of `require 'protocol_buffers` when compiled with your suggested codekitchen/ruby-protocol-buffers. 

I noticed that because I use the https://github.com/ruby-protobuf/protobuf compiler the complied .pb.rb files `require protocol_buffers`  are replaced with `require 'protobuf` instead.

This also means that (codekitchen) ` <::ProtocolBuffers::Message;`
is replaced by (ruby-protobuf) `< ::Protobuf::Message;`

I hope this gives you a clue of what could be wrong. Thanks again.
